### PR TITLE
Update to ProtoBuf 1.0.11

### PR DESCRIPTION
**Pull request by nomadbl**
This project is great, which is why I wanted to use it.
The problem was I was getting version conflicts with other libraries due to using a very old version of `ProtoBuf`
So, this PR does the following:
- Generate new code with ProtoBuf 1.0.11
- Modify the code as necessary to comply with the breaking changes in ProtoBuf. https://docs.juliahub.com/ProtoBuf/5SOUb/1.0.7/#Migrating-from-earlier-versions-of-ProtoBuf.jl
- update some code to use (new?) better syntax, such as not comparing with `nothing`, adding a bit of typing for clarity..
- Fix some tests (they assumed older Protobuf structs)

This update may break older Julia builds, so please help me test it as you see fit. I tested with Julia1.9 including running the tests and verifying the examples. Exceptions:
- Audio Example seems to not work (I don't hear the tone, maybe my speakers..)
- Graph feature needs some more editing. I marked with the line which breaks with `@test_broken`. I don't know what the internal graph functions do so please fix that in a following separate PR.
- A few examples are broken due to old code (i.e. MNIST code was updated since this was written, etc)

All in all, this update will make sure the project does not die when more libraries update.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
I don't have the time to maintain the package anymore, you should ping some other maintainers in the org...

**Comment by nomadbl** [comment]
@c42f @tanmaykm @quinnj @pfitzseb @fredrikekre 
Hi guys.
I would welcome a review by any of you that is willing.
I would also be willing to help maintaining this package if I get write permissions. I probably got a good overview of the package by working on this PR.

**Comment by PhilipVinc** [comment]
(BTW while you are at it, you should clean up the unit test runners to stop testing this on Julia 1.3, and probably bump the minimum required version to the 1.6 LTS..)

**Comment by oxinabox** [comment]
This looks good to me. I believe I have admin on the org and have been included in this package from the beginning.

I will try and give this a proper review early next week.

File: Project.toml [hunk 1]
```diff
@@ -14,10 +14,10 @@ StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
 [compat]
 FileIO = "1.2.3"
 ImageCore = "0.8.1, 0.9"
-ProtoBuf = "0.10, 0.11"
+ProtoBuf = "0.10, 0.11, 1.0.11"
```
Comment (oxinabox) [inline Project.toml]: Do we actually still support those old versions of ProtoBuf?
or just

```suggestion
ProtoBuf = "1.0.11"
```

File: Project.toml [hunk 2]
```diff
@@ -14,10 +14,10 @@ StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
...
 Requires = "0.5, 1"
-StatsBase = "0.27, 0.28, 0.29, 0.30, 0.31, 0.32, 0.33"
-julia = "1.3"
+StatsBase = "0.27, 0.28, 0.29, 0.30, 0.31, 0.32, 0.33, 0.34"
+julia = "1.3, 1.9"
```
Comment (oxinabox) [inline Project.toml]: This means the same as
```suggestion
julia = "1.3"
```
because `1.3` means `[1.3.0, 2.0)` (inclusive, exclusive bounds)


But if the oldest thing we are running CI in 1.6 then lets say
```suggestion
julia = "1.6"
```

File: gen/compile_proto.jl [hunk 3]
```diff
@@ -21,63 +21,47 @@ pbpath =dirname(dirname(PosixPath(pathof(ProtoBuf))))/p"gen"
 cur_path = cwd()
 TBL_root = dirname(cur_path)
 
-src_dir = cur_path/"proto"
-out_dir   = cur_path/"protojl"
+# src_dir = cur_path/"proto"
+src_dir = PosixPath(".")/"proto"
```
Comment (oxinabox) [inline gen/compile_proto.jl:25]: Why was this changed?
Doesn't this break things in windows?

File: gen/compile_proto.jl [hunk 4]
```diff
@@ -21,63 +21,47 @@ pbpath =dirname(dirname(PosixPath(pathof(ProtoBuf))))/p"gen"
...
-    # Include search paths
-    includes   = [src_dir, src_dir/base_module]
 
     # Output folder
-    module_out_dir    = out_dir/cur_module
+    module_out_dir = out_dir/cur_module
 
     # Input files
-    infiles = glob("*.proto", src_dir/input_path)
+    infiles = split.(string.(glob("*.proto", src_dir/input_path)), '/') .|> (a -> a[3:end]) .|> a -> joinpath(a...)
```
Comment (oxinabox) [inline gen/compile_proto.jl:38]: For readability can we avoid using `|>` ?

This inparticular I think is the same as:
```suggestion
    infiles = glob("proto/$(input_path)/*.proto", src_dir/input_path)), '/')
```

File: gen/compile_proto.jl [hunk 5]
```diff
@@ -21,63 +21,47 @@ pbpath =dirname(dirname(PosixPath(pathof(ProtoBuf))))/p"gen"
...
 
     mkpath(module_out_dir)
-    includes_str=["--proto_path=$path" for path=includes]
-    run(ProtoBuf.protoc(`$includes_str --julia_out=$module_out_dir $infiles`))
-
-    nothing
+    strip_currdir(str::String) = string(strip(str, ['.','/']))
```
Comment (oxinabox) [inline gen/compile_proto.jl]: is this required?

File: src/Deserialization/deserialization.jl [hunk 6]
```diff
@@ -137,19 +137,9 @@ end
...
-        return :audio
-    elseif hasproperty(summary, :tensor)
-        return :tensor
-    #elseif hasproperty(summary, :simple_value)
-    end
-    # always defined
-    return :simple_value
+function summary_type(summary::Summary_Value)
+    summary.value isa Nothing && error("Summary value of Nothing while deserializing")
+    return summary.value.name
```
Comment (oxinabox) [inline src/Deserialization/deserialization.jl:142]: Do we not need the others anymore?

File: src/Deserialization/deserialization.jl [hunk 7]
```diff
@@ -254,16 +243,18 @@ function map_summaries(fun::Function, logdir; purge=true, tags=nothing, steps=no
...
-
+            # if event.what contains no summary, don't bother processing this event
+            event.what isa Nothing && continue
+            !(event.what.value isa Summary) && continue
             step = event.step
             steps !== nothing && step ∉ steps && continue
 
-            iter = SummaryDeserializingIterator(event.summary, smart)
+            iter = SummaryDeserializingIterator(event.what.value, smart)
+            # fullnames = String[]
```
Comment (oxinabox) [inline src/Deserialization/deserialization.jl]:
```suggestion
```

File: src/Deserialization/deserialization.jl [hunk 8]
```diff
@@ -254,16 +243,18 @@ function map_summaries(fun::Function, logdir; purge=true, tags=nothing, steps=no
...
             for (name, val) in iter
                 tags !== nothing && name ∉ tags && continue
+                # fullname = "$(name)_$(step)"
+                # fullname in fullnames ? continue : push!(fullnames, fullname)
```
Comment (oxinabox) [inline src/Deserialization/deserialization.jl]:
```suggestion
```

File: src/Deserialization/images.jl [hunk 9]
```diff
@@ -14,7 +14,7 @@ function lookahead_deserialize_image_summary(old_tag, old_val, evs::Summary,
 
     state = state_old + 1
     iter_result = iterate(evs, state)
-    while iter_result !== nothing
+    while !(iter_result isa Nothing)
```
Comment (oxinabox) [inline src/Deserialization/images.jl]: The previous code was fine.
`!==` is that same as `!(iter_result === nothing)` 
which (unlike `!=`) the compiler can optimize easily based on type information.
Since `===` can only be true of things with same type, and since `Nothing` is a singleton type


See demonstration:
```julia
julia> foo(x) = x!==nothing
foo (generic function with 1 method)

julia> bar(x) = !(x isa Nothing)
bar (generic function with 1 method)

julia> @code_llvm foo(1)
;  @ REPL[1]:1 within `foo`
define i8 @julia_foo_133(i64 signext %0) #0 {
top:
  ret i8 1
}

julia> @code_llvm foo(nothing)
;  @ REPL[1]:1 within `foo`
define i8 @julia_foo_151() #0 {
top:
  ret i8 0
}

julia> @code_llvm bar(1)
;  @ REPL[2]:1 within `bar`
define i8 @julia_bar_156(i64 signext %0) #0 {
top:
  ret i8 1
}

julia> @code_llvm bar(nothing)
;  @ REPL[2]:1 within `bar`
define i8 @julia_bar_158() #0 {
top:
  ret i8 0
}
```
either way it constant folds

File: src/Optional/LogGraph.jl [hunk 10]
```diff
@@ -22,35 +23,59 @@ function graph_summary(g, nodelabel, nodeop, nodedevice, nodevalue)
...
+			attr["value"] = AttrValue(OneOf(:list, listvalue))
+			shape = TensorShapeProto([TensorShapeProto_Dim(length(x), "")], false)
+			listvalue = AttrValue_ListValue(Vector{Vector{UInt8}}(),
+											Vector{Int64}(),
+											Vector{Float32}(),
+											Vector{Bool}(),
+											Vector{var"#DataType".T}(), 
+											[shape],
+											Vector{TensorProto}(), 
+											Vector{NameAttrList}())
```
Comment (oxinabox) [inline src/Optional/LogGraph.jl]: Indenting seems to have gone very weird here.

File: test/runtests.jl [hunk 11]
```diff
@@ -11,8 +11,10 @@ ENV["GKSwstype"] = "100"
 ENV["DATADEPS_ALWAYS_ACCEPT"] = true
 
 LOG_DIRS = Any["test_logs/"]
+test_s3 = false # Minio breaks on my system
```
Comment (oxinabox) [inline test/runtests.jl:14]: should this be turned on on CI?

**Review by oxinabox — COMMENTED** [review]
Thanks, looks mostly good, can you take a look at my comments and we should be able to get this merged in short order

File: gen/compile_proto.jl (hunk 3 above)
Comment (nomadbl) [reply gen/compile_proto.jl:25]: Since this only runs **once** when building the package (specifically, when updating ProtoBuf) I don't think this should pose a problem. In particular there is nothing windows specific here (the generated files are commited and this has no reason to run under windows).

However if any Unit test fails under windows, then we can see about this.

Comment (nomadbl) [reply gen/compile_proto.jl:25]: In principle it should work either way but ProtoBuf 1.0.0 code generation is annoying, so if at all possible I would like to avoid repeating this.

File: gen/compile_proto.jl (hunk 5 above)
Comment (nomadbl) [inline gen/compile_proto.jl]:
```suggestion
```

Comment (nomadbl) [reply gen/compile_proto.jl]: This is indeed unused

File: gen/compile_proto.jl [hunk 12]
```diff
@@ -21,63 +21,47 @@ pbpath =dirname(dirname(PosixPath(pathof(ProtoBuf))))/p"gen"
...
+    # relative_paths = string.(infiles) .|> strip_currdir
```
Comment (nomadbl) [inline gen/compile_proto.jl]:
```suggestion
```

File: src/Deserialization/deserialization.jl (hunk 6 above)
Comment (nomadbl) [reply src/Deserialization/deserialization.jl:142]: Look carefully at the new generated structs, they are structured differently. This new version of the method is functionally identical to the last one, it just does it differently.

File: src/Deserialization/images.jl (hunk 9 above)
Comment (nomadbl) [reply src/Deserialization/images.jl]:
```suggestion
    while iter_result !== nothing
```

File: src/Optional/LogGraph.jl (hunk 10 above)
Comment (nomadbl) [reply src/Optional/LogGraph.jl]: Looks fine on my computer, must be some github formatting issue?

File: Project.toml (hunk 1 above)
Comment (nomadbl) [reply Project.toml]: Nope lol

File: gen/compile_proto.jl (hunk 4 above)
Comment (nomadbl) [reply gen/compile_proto.jl:38]: In all honesty, this code should probably avoid the use of `PosixPath` altogether, which I used here in the first place because it was already in use and I tried to change as little as possible.
As it is now, it's hard for me to say if this change will work or not. It worked when run, but fails in the REPL. In the interest of my sanity lets do this in a separate PR (which I'll be happy to help with).
Since this whole gen code only needs to run once, lets leave it for now :pray:

File: test/runtests.jl (hunk 11 above)
Comment (nomadbl) [reply test/runtests.jl:14]: As the comment suggests, this throws Minio errors on my machine which I was unable to fix, so I disabled it.
If you can run the tests with `true` here it would be a big help :+1: 
If it works for you, then sure lets turn that on for CI

File: src/Optional/LogGraph.jl (hunk 10 above)
Comment (oxinabox) [reply src/Optional/LogGraph.jl]: Oh it is using tabs for some reason.
Please reformat to use spaces, inline with the other files.

Comment (nomadbl) [reply src/Optional/LogGraph.jl]: I changed the tabs here to spaces. Is it any better now?

File: src/Loggers/LogHistograms.jl [hunk 13]
```diff
@@ -48,27 +48,36 @@ end
...
+    num = length(hist_vals)
+    histsum = sum(hist_vals)
+    histsumsqr = sum(hist_vals.^2) 
+    hp = HistogramProto(first(edges), last(edges),
+                        num,
+                        histsum,
+                        histsumsqr,
+                        collect(edges[2:end]),
+                        hist_vals)
+    return Summary_Value(name, name, nothing, OneOf(:histo, hp))
```
Comment (oxinabox) [inline src/Loggers/LogHistograms.jl:60]: Looks like this file got tabs too

Comment (nomadbl) [reply src/Loggers/LogHistograms.jl:60]: Actually no, but I checked all the files just now and found one that did, and fixed it.
